tortoise/tortoise.go

```diff
@@ -636,268 +449,264 @@ func (t *turtle) getBallotBeacon(ballot *types.Ballot, logger log.Log) ([]byte,
+				for _, bid := range t.blocks[lid] {
+					votes[bid] = against
+				}
+			}
+			for _, bid := range ballot.ForDiff {
+				votes[bid] = support
+			}
+			for _, bid := range ballot.NeutralDiff {
+				votes[bid] = abstain
+			}
+			for _, bid := range ballot.AgainstDiff {
+				votes[bid] = against
+			}
+			ballots = append(ballots, tortoiseBallot{
+				id:     ballot.ID(),
+				base:   ballot.BaseBallot,
+				weight: ballotWeight,
+				votes:  votes,
+			})
 		}
+	}
+	t.ballots[lid] = ballotsIDs
+	return ballots, nil
+}
 
-		logger.Info("attempting to verify candidate layer")
-		contextualValidity, err := t.verifyingTortoise(ctx, logger, candlid)
-		if err != nil {
+func (t *turtle) updateLocalOpinion(ctx *tcontext, from, to types.LayerID) error {
+	for lid := from; !lid.After(to); lid = lid.Add(1) {
+		if err := t.addLocalOpinion(ctx, lid, t.localOpinion); err != nil {
 			return err
 		}
+	}
+	return nil
+}
 
-		if contextualValidity != nil {
-			for blk, v := range contextualValidity {
-				if err := t.bdp.SaveContextualValidity(blk, candlid, v); err != nil {
-					logger.With().Error("error saving contextual validity on block", blk, log.Err(err))
-					return fmt.Errorf("saving contextual validity for %s: %w", blk, err)
-				}
+func (t *turtle) keepVotes(ballots []tortoiseBallot) {
+	for _, ballot := range ballots {
+		votes := ballot.votes
+		basevotes := t.full.votes[ballot.base]
+		if len(basevotes) == 0 {
+			t.logger.With().Warning("base votes are empty", ballot.id, log.Stringer("base_ballot", ballot.base))
+		}
+		for vote, sign := range basevotes {
+			votelid, exist := t.blockLayer[vote]
+			if !exist || !votelid.After(t.evicted) {
+				continue
 			}
-			t.Verified = candlid
-		} else {
-			healed, err := t.healingTortoise(ctx, logger, candlid)
-			if err != nil || !healed {
-				return err
+			if _, exist := votes[vote]; !exist {
+				votes[vote] = sign
 			}
 		}
-		logger.With().Info("verified candidate layer", log.Named("new_verified", t.Verified))
+		t.full.votes[ballot.id] = votes
 	}
-	return nil
+}
+
+// the idea here is to give enough room for verifying tortoise to complete. if verifying tortoise fails to verify
+// a layer within last hdist layers or within zdist + confidence param (whichever is larger) we switch to full vote counting.
+//
+// verifying tortoise is a fastpath for full vote counting, that works if everyone is honest and have good synchrony.
+// as for the safety and livenesss both protocols are identical.
+func (t *turtle) isHealingEnabled() bool {
+	lid := t.verified.Add(1)
+	return lid.Before(t.layerCutoff()) &&
+		t.last.Difference(lid) > t.Zdist &&
+		// unlike previous two parameters we count confidence interval from processed layer
+		// processed layer is significantly lower then the last layer during rerun.
+		// we need to wait some distance before switching from verifying to full during rerun
+		// as previous two parameters will be always true even if single layer is not verified.
```

Comment:
i don't think that works. the rerun is more likely to enter self-healing because we may learn new state (e.g. different votes and atxs)

Reply:
the intuition is that we need to have some heuristic to delay when we switch into full mode, during live tortoise this delay is hdist. during rerun, hdist is irrelevant as we are long past hdist. i want to discuss it on the next review with research. confidence param works for now, but maybe there are better solutions

Reply:
> i don't think that works. the rerun is more likely to enter self-healing because we may learn new state (e.g. different votes and atxs)

but it's just as easy for rerun to signal stuck-ness to tortoise and enter self-healing mode from there.
is it possible that verifying tortoise enter healing while rerun also enter healing?

self-healing however should get the full state from mesh for correctness?

Reply:
> but it's just as easy for rerun to signal stuck-ness to tortoise and enter self-healing mode from there.
is it possible that verifying tortoise enter healing while rerun also enter healing?

i am confused by this question. the comparison between verifying tortoise and rerun doesn't make sense. there is live tortoise (just a normal workflow) and rerun, and there is verifying tortoise and full tortoise. live tortoise can be in both modes and rerun can be in both modes.

> self-healing however should get the full state from mesh for correctness?

why? it has all the state it needs to make progress

Reply:
> is it possible that verifying tortoise enter healing while rerun also enter healing?

you meant live tortoise enters healing and later rerun enters full mode when it processes same layers? yes, in fact it is very likely to happen. for example if there were many ballots with incorrect beacon during live tortoise, and later during rerun verifying tortoise won't make progress as well

Reply:
yes. i meant that live tortoise entering full mode, and just a little bit later rerun enters full mode.

since full-mode is expensive and should be avoided, why not follow the heuristic.
when live tortoise gets stuck
- first wait for the latest rerun to finish (without entering full mode). if it made progress, swap rerun state.
- if the latest rerun also get stuck, then stop future reruns and enter full mode.
- resume live tortoise and rerun after exiting full mode

Reply:
> first wait for the latest rerun to finish (without entering full mode). if it made progress, swap rerun state.

i don't understand what is expected to get fixed by this. it is expected that full mode will fix our local opinion and eventually, we will be able to switch into verifying mode, either naturally or forcefully (as was discussed on last call https://github.com/spacemeshos/go-spacemesh/issues/2985)

> if the latest rerun also get stuck, then stop future reruns and enter full mode.

it is not full mode OR verifying tortoise & rerun. if we can't do rerun in verifying mode we will have to do it full mode, because of the same concerns, dishonest majority at the time of live-run and more uncounted weight from missing atxs.